x/protorev/protorev.md

```diff
@@ -0,0 +1,697 @@
+
+1. Check if the module is enabled.
+    1. If the module is disabled, nothing happens.
+2. Extract all pools that were traded on in the transaction (`ExtractSwappedPools`) as well as the direction of the trade.
+3. Create cyclic arbitrage routes for each of the swaps above (`BuildRoutes`)
+4. For each feasible route, determine if there is a cyclic arbitrage opportunity (`IterateRoutes`)
+    1. Determine the optimal amount to swap in and its respective profits via binary search over range of potential input amounts (`FindMaxProfitForRoute`)
+    2. Compare profits of each route, keep the best route and input amount with the highest profit
+5. If the best route and input amount has a profit > 0, execute the trade (`ExecuteTrade`) and rebalance the pools on-behalf of the chain through the `poolmanagerkeeper` (`MultiHopSwapExactAmountIn`)
+6. Keep the profits in the module’s account for subsequent distribution.
+
+### ExtractSwappedPools
+
+Checks if there were any swaps made on pools in a transaction, returning the pool ids and input/output denoms for each pool that was traded on.
+
+### BuildRoutes
+
+BuildRoutes takes a token pair (input and output denom) as well as the pool id and returns a list of routes for that token pair that potentially contain a cyclic arbitrage opportunity, populated via the Hot Route and Highest Liquidity Pools method as described above.
+
+### IterateRoutes
+
+IterateRoutes iterates through a list of routes, determining the route and input amount that results in the highest cyclic arbitrage profits..
+
+### FindMaxProfitForRoute
+
+This will take in a route and determine the optimal amount to swap in to maximize profits, given the reserves of all of the pools that are swapped against in the route.
+
+### ExecuteTrade
+
+Execute trade takes the route and optimal input amount as params, mints the optimal amount of input coin, executes the swaps via `poolmanagerKeeper`’s `MultiHopSwapExactAmountIn`, and then burns the amount of coins originally minted, storing the profits in it’s own module account.
+
+This will also update various trading statistics in the module’s store. It will update the total number of trades the module has executed, total profits captured, profits made on this specific route, share of profits the developer account can withdraw, and mor.
+
+## Execution Guardrails
+
+`x/protorev` is bounded and limited in the number of trades the module can execute per block. The purpose of doing so is to ensure that the current block time does not substantially change and that the module does not introduce a new attack vector. 
+
+Execution is currently limited in the following ways
+
+1. The binary search method for finding input amounts is bounded by some number of iterations.
+2. The number of routes that can be traversed in a given transaction is bounded by some number.
+3. The number of routes that can be traversed in a given block is bounded by some number.
```

Comment:
Asking to learn: how are these numbers decided on?

Reply:
@NotJeremyLiu

Reply:
> LGTM just one question
> 
> With stepSide inside route proto, do we check to see **each** route matches the stepSize amount now?

Yup! The step size now matches directly with each route.

Reply:
TLDR: We choose some number of iterations to bound by doing benchmark testing.